heatmap/src/heatmap.rs

```diff
@@ -280,52 +308,43 @@ impl Heatmap {
+                    let mut ticks_forward = elapsed.as_nanos() / self.resolution.as_nanos();
 
-                    // now we have a slice to subtract and clear from the summary
-                    // this is the histogram that is one ahead of our new current
-                    // position
-                    let mut to_clear = next + 1;
-
-                    // check if we need to wrap around to the start
-                    if to_clear >= self.slices.len() {
-                        to_clear -= self.slices.len();
+                    // move current and next_tick forward
+                    curr_tick += Duration::from_nanos(self.resolution.as_nanos() * ticks_forward);
+                    next_tick = curr_tick + self.resolution;
+
+                    self.curr_tick.store(curr_tick, Ordering::Relaxed);
+                    self.next_tick.store(next_tick, Ordering::Relaxed);
+
+                    // Note that in steady state, the next slice of the previous "current"
+                    // histogram (which was marked by `idx`) is the oldest slice of histogram
+                    // and therefore needs to be cleaned up. Similarly, if we need to move
+                    // the tick forward by more than one window, all histograms we encounter
+                    // while advancing the `idx` will need to be cleaned up. Of course, here
+                    // the index will be wrapped around the ring buffer if it gets to the end.
+                    let mut idx = self.idx.load(Ordering::Relaxed) as usize;
+                    while ticks_forward > 0 {
+                        idx += 1;
+                        if idx == self.histograms.len() {
+                            idx = 0;
+                        }
+                        let _ = self.summary.subtract_and_clear(&self.histograms[idx]);
+
+                        ticks_forward -= 1
                     }
 
-                    self.start.fetch_add(self.resolution, Ordering::Relaxed);
-                    self.stop.fetch_add(self.resolution, Ordering::Relaxed);
-
-                    // check to make sure it's actually an old slice. A newly
-                    // created heatmap will have all slices in the future.
-                    if self.slices[to_clear].start.load(Ordering::Relaxed)
-                        < self.start.load(Ordering::Relaxed)
-                    {
-                        // subtract and clear
-                        let _ = self
-                            .summary
-                            .subtract_and_clear(&self.slices[to_clear].histogram);
-
-                        self.slices[to_clear]
-                            .start
-                            .fetch_add(self.span, Ordering::Relaxed);
-                        self.slices[to_clear]
-                            .stop
-                            .fetch_add(self.span, Ordering::Relaxed);
-                    }
+                    self.idx.store(idx, Ordering::Relaxed);
```

Comment:
Since we've moved `next_tick` forward, but not `idx` won't we record into the wrong slice while we're cleaning up old slices?

Reply:
Yea I totally forgot the lock only for `tick()` part. tentatively made some changes but I need to go back and read the various memory ordering again.

Reply:
this is no longer applicable after the latest change.